### PR TITLE
Input handling v2

### PR DESCRIPTION
**Pull request by tma-isbx**
This would be a second generation version of my previous input handling pull request (#166) rewritten from "scratch".

Primary item of interest is that it now compiles the key mappings into a bennu (https://github.com/mattbierner/bennu) based incremental parser and uses that for parsing the input stream.

Motion key mappings are defined like:
* `<count>w`
* `<count>t<argument>`
* `0`
* `gg`
* `<count>G`

Command key mappings are defined like:
* `<register><count>yy`
* `<register><count>y<range>`
* `m<argument>`
* `<register><count>dd`
* `<esc>`
* `<c-f>`

`<register>` is an optional sequence of `"<argument>`
`<count>` is an optional sequence of [1-9][0-9]* (special allowance for the 0 command)
`<argument>` is a single key
`<range>` = `<count><motion>` or `<textObject>`

I've started moving keys to be `<esc>` instead of just "esc" because I needed to differentiate between the key stroke `<esc>` versus the keys e-s-c typed in sequence.

As with my other pull request, mode switching should again probably be moved out of modeHandler.handleKeyEvent() as it intercepts i/a and prevents the user from doing "fa".

### CODE REVIEW

File: src/mode/mode.ts [hunk 1]
```diff
@@ -42,35 +49,97 @@ export abstract class Mode {
...
-        "w" : async (c) => { return c.getWordRight(); },
-        "e" : async (c) => { return c.getCurrentWordEnd(); },
-        "b" : async (c) => { return c.getWordLeft(); },
-        "}" : async (c) => { return c.getCurrentParagraphEnd(); },
-        "{" : async (c) => { return c.getCurrentParagraphBeginning(); }
+
+        "<count>t<argument>" : async (p, o) => { return p.tilForwards(o.argument, o.count); },
+        "<count>T<argument>" : async (p, o) => { return p.tilBackwards(o.argument, o.count); },
+        "<count>f<argument>" : async (p, o) => { return p.findForwards(o.argument, o.count); },
+        "<count>F<argument>" : async (p, o) => { return p.findBackwards(o.argument, o.count); }
```
Comment (bradphelan) [inline src/mode/mode.ts]: Great : This pull request implements https://github.com/VSCodeVim/Vim/issues/187

**Comment by bradphelan** [comment]
I couldn't get the branch to build. When running gulp it would always say that the "bennu" module was missing. From my limited experience with typescript it seems it is missing the typings file as bennu is pure javascript. Did you forget to check this in?

**Comment by tma-isbx** [comment]
I am not aware of typings file for bennu but it may exist. However, the Typescript -> Javascript compilation still occurs, and the Javascript that is ultimately loaded runs without issue.

**Comment by bradphelan** [comment]
It doesn't for me. I've the latest node, npm and gulp. And it tells me that
benny can't be found. The bennu library is there in the node modules
directory but the typescript import command fails to load it.

**Comment by tma-isbx** [comment]
This is what happens for me:

~~~
# delete my output directory
vscode-vim> rm -rf out

# compile, get errors
vscode-vim> npm run compile | grep bennu
src/mode/keyParser.ts(4,42): error TS2307: Cannot find module 'bennu'.
^C

# output is still generated
vscode-vim> ls out/out/src/mode/keyParser.js
out/src/mode/keyParser.js
~~~

**Comment by bradphelan** [comment]
Ok you are right. It seems errors are not errors when dealing with typescript compilation :) I'm too used of c#.  I can conform that __fi__ and __fa__ are not working because they are grabbed by the __i__ and __a__ 

Nice work though. I'm looking forward for this to be folded into the main branch.

**Comment by jpoon** [comment]
If you rebase to master, that should resolve the build issues. WHen do you think this will be ready for review/merge?

**Comment by tma-isbx** [comment]
I should be able to resolve the conflicts and add unit tests in the next day or two. I may leave the mode switching changes to another pull request.

**Comment by jpoon** [comment]
Cool. I ask as I'm in somewhat of a holding pattern as any further work I do will make merge conflicts a nightmare.

**Comment by tma-isbx** [comment]
I've got most of the changes done, just fixing up a unit test.

**Comment by tma-isbx** [comment]
Everything should be merged, and tests passing.

`gulp tslint` does complain about a use before declaration (haven't figured out quite how I'll address that or if I really need to) and `npm run compile` also has a couple of additional "errors" (I assume due to modules without .td files, and some Javascript-isms).

**Comment by jpoon** [comment]
@tma-isbx. Tests don't seem to be running against your PR? Not sure why, but I don't even see it on Travis.

Any chance we can get around those compile errors? Can we maybe create our own typescript definition -- https://github.com/Microsoft/TypeScript-Handbook/blob/master/pages/Writing%20Definition%20Files.md

**Comment by tma-isbx** [comment]
Not sure about Travis. Tests seem to run locally within VSCode and also `npm run test`.

I can look into creating my own typescript definitions. I'll be out this weekend, so it will probably be a couple of days.

**Comment by tma-isbx** [comment]
Updated to address compile errors.

**Comment by aakashsigdel** [comment]
Any updates on this one?

**Comment by jpoon** [comment]
I'm going to close this PR due to inactivity. As it stands, it's also wayyyy out of date with master. Feel free to re-open.

**Comment by johnfn** [comment]
@jpoon & others: I'm gonna attempt to bring this one up to date with master and merge it in. I gotta digest everything that's going on here first, though.

**Comment by tma-isbx** [comment]
Apologies, I thought I had addressed everything at the time of my last update. Unfortunately I've been very busy since (and currently), I don't believe I have time in the near future to bring it up to date to master.

**Comment by johnfn** [comment]
OH god, I take it back, there's no way I can merge this in now, so much has changed in master! I now understand what you were trying to do though @tma-isbx, so I hope to take that inspiration going forward.

**Pull request closed without merging**